### DISCUSSION TITLE
istioctl proxy-status and Kiali not working because port forwarding not working?

### DISCUSSION THREAD

**Discussion by dmosesson** [istio/istio#48681] [Q&A]
I have a cluster where I installed istio, and it seems to work. I am trying to get kiali installed so that I can see all the things that istio is doing, and kiali is showing that `istiod-full-pod-name` is not reachable.

Looking into the logs, the error seems to be: 
`portforward.go:409] and error occurred forwarding 14723 -> 8080: error forwarding port to pod ...`
Failed to call Istiod endpoint /debug/syncz error: unable to proxy Istiod pods. Make sure your Kubernetes API server has access to the Istio control plane through 8080 port`

I know that in my cluster, for security reasons, kubectl port-forward doesn't work. Is there a way to set up kiali to hit the istiod endpoint without needing port forwarding? They are in the same namespace, and and the 8080 port on the pod seems to reachable just fine directly. I get basically the same error from `istioctl proxy-status` if that helps anything.

Is there anything I can do so that kaili/istioctl just hit services directly? I'm happy to create a deployment for istioctl inside the cluster if that would help.

**Comment by husnialhamdani** [comment]
Maybe you can enable the ingress endpoint https://kiali.io/docs/installation/installation-guide/accessing-kiali/#ingress-access

**Reply by dmosesson** [reply]
I think I may not have been clear.

Kiali itself is accessible through an ingress, but it cannot reach istiod in the same cluster, because apparently (at least for the way I have it set up) it is using port forwarding to make that connection happening. Tentatively, that is happening here: https://github.com/kiali/kiali/blob/edafb9234b5015ed41fb1b0626c78a12730b2508/util/httputil/port_forwarder.go#L30

**Comment by hanxiaop** [comment]
Have you tried `istioctl x ps`, which is based on xDS instead of port-forward.

**Reply by dmosesson** [reply]
I'll try that when I get in tomorrow. Is there an equivalent for Kiali?

The closest I see is https://github.com/kiali/kiali/pull/5533, which as the last comment on that PR points out, is not documented.

**Reply by hanxiaop** [reply]
AFAIK they are using the `/debug` endpoint, you may ask if there are workarounds in Kiali repo.

**Reply by dmosesson** [reply]
`istioctl x ps` seems to be using port forwarding as well. If this is different between 1.19.3 and and 1.20.1 I am happy to upgrade but I don't see anything about it in the release notes

**Reply by hanxiaop** [reply]
It's not port forwarding but connecting to your istiod xds address. You may check if you have access to the xds address, and see the related flags with `istioctl x ps -h`

**Reply by dmosesson** [reply]
It still uses port forwarding at least in part for that

**Reply by hanxiaop** [reply]
I mean you can use `istioctl x ps --xds-address <your istiod xds address>` to directly make request to the xds server without port-forwarding.
